packages/api/src/Base.ts

```diff
@@ -338,122 +339,115 @@ export default abstract class ApiBase<OnCall> implements ApiBaseInterface<OnCall
     return output;
   }
 
-  private decorateRpc<T> (
-    rpc: RpcRx,
-    onCall: (method: OnCallFunction<Observable<Codec | undefined | null>>, params: Array<any>, isSubscription?: boolean) => T
-  ): DecoratedRpc<T> {
-
+  private decorateRpc<C, S> (rpc: RpcRx, onCall: OnCallDefinition<C, S>): DecoratedRpc<C, S> {
     return ['author', 'chain', 'state', 'system'].reduce((result, _sectionName) => {
-      const sectionName = _sectionName as keyof DecoratedRpc<T>;
+      const sectionName = _sectionName as keyof DecoratedRpc<C, S>;
 
       result[sectionName] = Object.keys(rpc[sectionName]).reduce((section, methodName) => {
         // FIXME Find a better way to know if a particular method is a subscription or not
-        const isSubscription = methodName.includes('subscribe');
-        const method = (...params: any[]) => onCall(rpc[sectionName][methodName], params, isSubscription);
+        // const isSubscription = methodName.includes('subscribe');
```

Comment:
if you remove the isSubscription information,`apiPromise.rpc.state.subscribeStorage(key)` will return a Promise.

Or are we waiting for the isSubscription bubbling from rpc-core?

Reply:
Ummm, you are right.

We actually need a `needsSubscription` flag here. i.e. if set and callback is not provided, throw.

Reply:
Ok, added a `needsCallback` that does a check in `onCall` (promise), at least we have a runtime error.

Along with this, update RPC to have proper-ish params as well. Derive is till missing, it is all any, any, any